Rehome CCS/CCR permissions (#6934)
elastic/docs-content: explore-analyze/cross-cluster-search.md

```console
POST /_security/user/cross-search-user
{
  "password" : "l0ng-r4nd0m-p@ssw0rd",
  "roles" : [ "remote-search" ]
}
```

:::{note}
You only need to create this user and role on the **local** cluster.

The same user can hold multiple roles, or a single role can combine [remote indices privileges](/deploy-manage/users-roles/cluster-or-deployment-auth/role-structure.md#roles-remote-indices-priv) with local index privileges and {{kib}} access roles.
:::

### TLS certificate authentication [configure-privileges-for-ccs-cert]
```{applies_to}
stack: deprecated 9.0
```

:::{warning}

Certificate based authentication is deprecated. Configure [API key authentication](/deploy-manage/remote-clusters/remote-clusters-api-key.md) instead or follow a guide on how to [migrate remote clusters from certificate to API key authentication](/deploy-manage/remote-clusters/remote-clusters-migrate.md).
:::

> After [connecting remote clusters](/deploy-manage/remote-clusters/remote-clusters-self-managed.md), create matching user roles on both the local and remote clusters and assign the necessary privileges. With TLS-based authentication, the local user's role names are forwarded to the remote cluster, which authorizes the request by evaluating roles with the same names defined locally.

Reviewer: I'd consider whacking all this in a dropdown to underline the "are you sure?" messaging and slimming the page down a bit :)

Author: Love this! If it's deprecated it should be prioritised less. 👍 

I'll add this suggestion to https://github.com/elastic/docs-content/issues/7064 as there's a few more improvements that could be applied to some of the sections. I'll implement this as part of 7064

Assuming the remote cluster is connected under the name of `my_remote_cluster`, the following request creates a `remote-search` role on the local cluster that allows searching the remote `target-index` index:

```console
POST /_security/role/remote-search
{
  "remote_indices": [
    {
      "clusters": [ "my_remote_cluster" ],
      "names": [
        "target-index"
      ],
      "privileges": [
        "read",
        "read_cross_cluster",
        "view_index_metadata"
      ]
    }
  ]
}
```

After creating the `remote-search` role, use the [create or update users]({{es-apis}}operation/operation-security-put-user) API to create a user on the local cluster and assign the `remote-search` role. For example, the following request assigns the `remote-search` role to a user named `cross-search-user`:

```console
POST /_security/user/cross-search-user
{
  "password" : "l0ng-r4nd0m-p@ssw0rd",
  "roles" : [ "remote-search" ]
}
```

:::{note}
You only need to create this user and role on the **local** cluster.

> The same user can hold multiple roles, or a single role can combine [remote indices privileges](/deploy-manage/users-roles/cluster-or-deployment-auth/role-structure.md#roles-remote-indices-priv) with local index privileges and {{kib}} access roles.

Reviewer 2: I like that line a lot! It's not in the scope of this doc to show how to grant Kibana access to a user.

:::

The following request creates a `remote-search` role on the remote cluster:

```console
POST /_security/role/remote-search
{
  "indices": [
    {
      "names": [
        "target-indices"
      ],
      "privileges": [
        "read",
        "read_cross_cluster"
      ]
    }
  ]
}
```

#### Local cluster [configure-privileges-for-ccs-cert-local]

On the local cluster, which is the cluster used to initiate cross cluster search, assign users the `remote-search` role. If users only need remote access, you can leave the local role empty. If they also need to query local indices or use {{kib}}, grant the required local privileges in the same role or assign the user additional roles.

The following request creates a `remote-search` role on the local cluster with no privileges:

```console
POST /_security/role/remote-search
{}
```

:::{note}
In {{kib}}, you query remote indices using the `<remote_cluster_name>:<target>` pattern. To give users local index or {{kib}} access, add the required privileges to `remote-search` on the local cluster, or assign them separate roles.
:::

> 

Reviewer 2: ```suggestion
```

I think that note doesn't add any extra benefit to this section (TLS based cert authentication).

We are saying already this in the previous paragraph:

> If they also need to query local indices or use {{kib}}, grant the required local privileges in the same role or assign the user additional roles.

The previous is enough IMO, and on the other hand, CCS always queries remote data using the `<remote_cluster_name>:<target>`, it's not specific to this section, and doesn't have implications in the roles definition anymore.

Author: Removed! thank you!

```console
POST /_security/role/remote-search
{
  "indices": [
    {
      "names": [
        "target-indices"
      ],
      "privileges": [
        "read",
        "read_cross_cluster"
      ]
    }
  ]
}
```

#### Local cluster [configure-privileges-for-ccs-cert-local]

On the local cluster, which is the cluster used to initiate cross cluster search, a user only needs the `remote-search` role. The role privileges can be empty.

The following request creates a `remote-search` role on the local cluster:

```console
POST /_security/role/remote-search
{}
```

After creating the `remote-search` role on each cluster, use the [create or update users]({{es-apis}}operation/operation-security-put-user) API to create a user on the local cluster and assign the `remote-search` role. For example, the following request assigns the `remote-search` role to a user named `cross-search-user`:

```console
POST /_security/user/cross-search-user
{
  "password" : "l0ng-r4nd0m-p@ssw0rd",
  "roles" : [ "remote-search" ]
}
```

::::{note}
You only need to create this user on the **local** cluster.
::::

#### {{kib}} users [configure-privileges-for-ccs-kibana-cert]

> When using {{kib}} to search across multiple clusters, a two-step authorization process determines whether or not the user can access data streams and indices on a remote cluster:

Author: fixed in 8132bb348d9dc140037575778659dece6757e441

To grant users read access on the remote data streams and indices, you must create a matching role on the remote clusters that grants the `read_cross_cluster` privilege with access to the appropriate data streams and indices.

For example, you might be actively indexing {{ls}} data on a local cluster and periodically offload older time-based indices to an archive on your remote cluster. You want to search across both clusters, so you must enable {{kib}} users on both clusters.

On the local cluster, create a `logstash-reader` role that grants `read` and `view_index_metadata` privileges on the local `logstash-*` indices.

::::{note}
If you configure the local cluster as another remote in {{es}}, the `logstash-reader` role on your local cluster also needs to grant the `read_cross_cluster` privilege.
::::

```console
POST /_security/role/logstash-reader
{
  "indices": [
    {
      "names": [
        "logstash-*"
        ],
        "privileges": [
          "read",
          "view_index_metadata"
          ]
    }
  ]
}
```

> Assign your {{kib}} users a role that grants [access to {{kib}}](elasticsearch://reference/elasticsearch/roles.md), as well as your `logstash_reader` role. For example, the following request creates the `cross-cluster-kibana` user and assigns the `kibana-access` and `logstash-reader` roles.

Author: Fixed in a separate commit (8132bb348d9dc140037575778659dece6757e441)

* If you use [proxy mode](/deploy-manage/remote-clusters/remote-clusters-self-managed.md#proxy-mode), the local coordinating node must be able to connect to the configured `proxy_address`. The proxy at this address must be able to route connections to gateway and coordinating nodes on the remote cluster.
* {{ccs-cap}} requires different security privileges on the local cluster and remote cluster. Refer to [Configure privileges](#configure-privileges-for-ccs) for details.

> ## Configure privileges for {{ccs}} [configure-privileges-for-ccs]

Reviewer 2: All the intro in this section could be in a snippet, for your consideration.

After [remote clusters are connected](/deploy-manage/remote-clusters.md), you can configure which users on your local cluster can search data on remote clusters. The steps depend on the [remote cluster security model](/deploy-manage/remote-clusters/security-models.md) in use:

> * [API key authentication](#configure-privileges-for-ccs-api-key) (recommended), where you create roles with the required privileges on the local cluster.

Reviewer 2: ```suggestion
* [API key authentication](#configure-privileges-for-ccs-api-key) (recommended), where you create roles with the required remote privileges on the local cluster.
```

You can manage roles in {{kib}} on the **Roles** page in the navigation menu or use the [global search field](/explore-analyze/find-and-organize/find-apps-and-objects.md). You can also use the [role management]({{es-apis}}group/endpoint-security) APIs to add, update, remove, and retrieve roles dynamically. When you use the UI or APIs to manage roles, the roles are stored in an internal {{es}} index. When you use local files, the roles are only stored in those files. For more information, refer to [Defining roles](/deploy-manage/users-roles/cluster-or-deployment-auth/defining-roles.md).

The following examples use the [create or update roles]({{es-apis}}operation/operation-security-put-role) API and the [create or update users]({{es-apis}}operation/operation-security-put-user) API. You must have at least the `manage_security` [cluster privilege](elasticsearch://reference/elasticsearch/security-privileges.md#privileges-list-cluster) to use these APIs.

### API key authentication [configure-privileges-for-ccs-api-key]

Authorization for {{ccs}} works in two parts:

> * The [cross-cluster API key](/deploy-manage/remote-clusters/remote-clusters-api-key.md) used to connect to a remote cluster defines the maximum access that cluster allows.

Reviewer 2: ```suggestion
* The [cross-cluster API key](/deploy-manage/remote-clusters/remote-clusters-api-key.md) used to connect to a remote cluster defines the maximum access that the remote cluster cluster allows to any user. This key is created and configured during the remote cluster connectivity set up.
```

For your consideration, maybe we can remove the last part, or mention that the creation of it it's out of the scope of this document.

Author: Sounds good, adding this into the snippet I'm reusing.

> * Roles on the local cluster with [remote indices privileges](/deploy-manage/users-roles/cluster-or-deployment-auth/role-structure.md#roles-remote-indices-priv) or [remote cluster privileges](/deploy-manage/users-roles/cluster-or-deployment-auth/role-structure.md#roles-remote-cluster-priv) further limit which remote indices each user can search.

Reviewer 2: That sentence is incorrect in the way that it could lead the reader to think that a user without a local role would have `ALL` privileges that the API key used to interconnect clusters has. And that's not true.

The role on the local cluster can GRANT remote privileges to specific users, and the effective permissions are the interesction of the local roles + the API key used as a boundary.

Author: Perfect, I'm updating it to say:

> By default, users have no remote privileges unless they are superusers or are assigned a role that includes remote privileges. A user's effective access is the intersection of their role privileges and the API key privileges.

> * Roles on the local cluster with [remote indices privileges](/deploy-manage/users-roles/cluster-or-deployment-auth/role-structure.md#roles-remote-indices-priv) or [remote cluster privileges](/deploy-manage/users-roles/cluster-or-deployment-auth/role-structure.md#roles-remote-cluster-priv) further limit which remote indices each user can search.

Reviewer 2: ```suggestion
* Roles on the local cluster with [remote indices privileges](/deploy-manage/users-roles/cluster-or-deployment-auth/role-structure.md#roles-remote-indices-priv) or [remote cluster privileges](/deploy-manage/users-roles/cluster-or-deployment-auth/role-structure.md#roles-remote-cluster-priv) grant access to specific remote resources for individual users. A user's effective permissions are the intersection of these role privileges and the permissions allowed by the API key configured for the remote cluster connection.
```

We could add that `superusers` on the local clusters will have all remote privileges by default, allowing them to access the remote cluster resources defined by the API key used during the setup, if we consider that useful.

Author: Yes, I think it's an important detail to mention, thank you. It'll be included in the new snippet.

::::{note}
The cross-cluster API key used by the local cluster to connect the remote cluster must have sufficient privileges to cover all remote indices privileges required by individual users.
::::

Assuming the remote cluster is connected under the name of `my_remote_cluster`, the following request creates a `remote-search` role on the local cluster that allows searching the remote `target-index` index:

```console
POST /_security/role/remote-search
{
  "remote_indices": [
    {
      "clusters": [ "my_remote_cluster" ],
      "names": [
        "target-index"
      ],
      "privileges": [
        "read",
        "read_cross_cluster",
        "view_index_metadata"
      ]
    }
  ]
}
```

After creating the `remote-search` role, use the [create or update users]({{es-apis}}operation/operation-security-put-user) API to create a user on the local cluster and assign the `remote-search` role. For example, the following request assigns the `remote-search` role to a user named `cross-search-user`:

```console
POST /_security/user/cross-search-user
{
  "password" : "l0ng-r4nd0m-p@ssw0rd",
  "roles" : [ "remote-search" ]
}
```

::::{note}

> You only need to create this user on the **local** cluster.

Reviewer 2: ```suggestion
You only need to create this user and role on the **local** cluster.
```

::::

### TLS certificate authentication [configure-privileges-for-ccs-cert]
```{applies_to}
stack: deprecated 9.0
```

::::{warning}

Certificate based authentication is deprecated. Configure [API key authentication](/deploy-manage/remote-clusters/remote-clusters-api-key.md) instead or follow a guide on how to [migrate remote clusters from certificate to API key authentication](/deploy-manage/remote-clusters/remote-clusters-migrate.md).
::::

> After [connecting remote clusters](/deploy-manage/remote-clusters/remote-clusters-self-managed.md), create a user role on both the local and remote clusters and assign the necessary privileges.

Reviewer 2: ```suggestion
After [connecting remote clusters](/deploy-manage/remote-clusters/remote-clusters-self-managed.md), create matching user roles on both the local and remote clusters and assign the necessary privileges. With TLS-based authentication, the local user's role names are forwarded to the remote cluster, which authorizes the request by evaluating roles with the same names defined locally.```

Reviewer 2: For your evaluation, maybe it's not worthy to give this level of detail.
One thing that would be great is to transmit the message that maintaining and configuring privileges on this security model (tls based auth) is a bit of a mess and much more difficult to maintain than the new security model.

If a local user has 10 roles assigned, the names could potentially match extra roles in the remote cluster, granting privileges that maybe the administrator didn't want.

That's why I consider useful at least mentioning how this work: " the local user's role names are forwarded to the remote cluster, which authorizes the request by evaluating roles with the same names defined locally."

Author: OK that makes sense to me, adding in a new commit.

To grant users read access on the remote data streams and indices, you must create a matching role on the remote clusters that grants the `read_cross_cluster` privilege with access to the appropriate data streams and indices.

For example, you might be actively indexing {{ls}} data on a local cluster and periodically offload older time-based indices to an archive on your remote cluster. You want to search across both clusters, so you must enable {{kib}} users on both clusters.

On the local cluster, create a `logstash-reader` role that grants `read` and `view_index_metadata` privileges on the local `logstash-*` indices.

::::{note}
If you configure the local cluster as another remote in {{es}}, the `logstash-reader` role on your local cluster also needs to grant the `read_cross_cluster` privilege.
::::

```console
POST /_security/role/logstash-reader
{
  "indices": [
    {
      "names": [
        "logstash-*"
        ],
        "privileges": [
          "read",
          "view_index_metadata"
          ]
    }
  ]

> }

Reviewer 2: Kibana users section feels misleading at first sight! (I know this comes from the old Elasticsearch doc, but it's super old).

It makes the reader to think those users could be different to the users mentioned previously, or that Kibana changes any of this in the way it works, when it doesn't!

We need to clarify the scope in the introduction better, **if this is needed at all**. The only benefit this section gives is to tell users that they can "REUSE" the role to grant specific privileges on the local cluster (like kibana access + indices read on the local cluster), and in the previous example we used an EMPTY ROLE in the local cluster. And that's pretty obvious and could be covered by a note or something.

^^ That's the only benefit / use case, so I wouldn't create all this "kibana users section" under the deprecated TLS cert-based authentication and instead integrate that possibility in the previous block (actually the user could create an extra role for local resources instead of doing this "hack" but anyway).

* Also note that the concept that section covers is also applicable to the API key based auth, they could reuse the role to give local privileges.

* Also note that the order of the role creation is different here: we start with the local role and we end up with the remote cluster role.

Conclussions:

* I think it's time to simplify this by removing this entire section and just enrich the previous ones.

* We could include a similar concept to the API-key based authentication: it's a bit obvious but we can remind readers that the user on the local cluster could also have privileges to the local cluster indices and resources with other roles (or the same role) (obvious IMO and the only purpose of this section).

Author: Okay, I think I get it. I've made changes in 42ad41e7396c5a77852082788e8421a76774586a that hopefully address this.
